### PR TITLE
feat: readHomeDirFiles

### LINKED ISSUE #10534: Support reading credentials from home directory
**What would you like Renovate to be able to do?**

Allow Renovate to read credentials files such as `.npmrc`, `settings.xml` and `nuget.config` from home directories.

**Did you already have any implementation ideas?**

The reason for doing this is because it's what a lot of users expect or know, so the goal is to make it simpler for them even if it's more complex for us.

First of all, these files should not be read by Renovate unless enabled by an admin config option, e.g. something new named `homeDirConfig` which defaults to `false`. For best UX, we should check if any of these files exist at startup and log an INFO message if we are *ignoring* them.

The files should be read once at startup and converted into `hostRules` and maybe `registryUrls`/`packageRules` if necessary. They should be added *before* any explicitly configured rules, so that explicit config takes precedence.

The files will *not* be mapped to sidecar containers if `binarySource=docker`. We should document which file types are supported and which fields are extracted.

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Adds new option `readHomeDirFiles`.

## Context:

Closes #10534 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/global/config/parse/home-dir.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import os from 'os';
+import fs from 'fs-extra';
```
Comment (viceice) [inline lib/workers/global/config/parse/home-dir.ts:2]: Better use our fs proxies?

Comment (rarkins) [reply lib/workers/global/config/parse/home-dir.ts:2]: I couldn't get these tests to work any way I tried. Any ideas?

Comment (viceice) [reply lib/workers/global/config/parse/home-dir.ts:2]: I'll try to play with tomorrow.

**Review by viceice — APPROVED** [review]
got it working 🎉

**Comment by rarkins** [comment]
Thanks :)

File: lib/workers/global/config/parse/home-dir.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+          authType: 'Basic',
+        });
+      }
+      for (const [key, val] of Object.entries(npmrc)) {
+        const [scope, topic] = key.split(':');
+        if (topic === 'registry') {
+          config.packageRules.push({
+            matchDatasources: ['npm'],
+            matchPackagePrefixes: [scope],
+            registryUrls: [val],
```
Comment (viceice) [inline lib/workers/global/config/parse/home-dir.ts:50]: I think this is currently not used by npm datasource?

File: lib/workers/global/config/parse/home-dir.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+          });
+        }
+        if (topic === '_authToken') {
+          const matchHost = scope.replace(/^\/\//, 'https://');
+          config.hostRules.push({ matchHost, token: val });
+        }
+        if (topic === '_auth') {
```
Comment (viceice) [inline lib/workers/global/config/parse/home-dir.ts:57]: i think there is some legacy `username` and `_password`  option too 🤔

File: lib/workers/global/config/parse/home-dir.ts (hunk 2 above)
Comment (rarkins) [reply lib/workers/global/config/parse/home-dir.ts:50]: Yes, you're right. Thinking if I could/should change that though. I can't think what held it back before.

File: lib/workers/global/config/parse/home-dir.ts (hunk 3 above)
Comment (rarkins) [reply lib/workers/global/config/parse/home-dir.ts:57]: I think you're right, based on this: https://docs.microsoft.com/en-us/azure/devops/artifacts/npm/npmrc?view=azure-devops#credentials-setup

**Comment by rarkins** [comment]
I'm going to rewrite how this works

**Pull request closed without merging**